### DISCUSSION TITLE
About `:matches-property()` pseudo-class

### DISCUSSION THREAD

**Discussion by superlex** [uBlockOrigin/uBlock-issues#2981] [Interoperability] [answered]
Hi,
AdGuard supports `:matches-property()` pseudo-class:
https://adguard.com/kb/it/general/ad-filtering/create-own-filters/#extended-css-property 

Does uBlock support it, or it may support it in the future?

Currently some rules in AG filters lists for uBO contain it, and they are invalid.

Example:
`fotor.com#$?##main > div > div + div:matches-property(offsetWidth=160) { remove: true; }`
found in https://filters.adtidy.org/extension/ublock/filters/2_without_easylist.txt

**Comment by MasterKia** [comment]
`:matches-attr()`:
https://github.com/uBlockOrigin/uBlock-issues/wiki/Procedural-cosmetic-filters#subjectmatches-attrarg

**Reply by mapx-** [reply]
maybe `matches-property` is more like uBO's [matches-media](https://github.com/uBlockOrigin/uBlock-issues/wiki/Procedural-cosmetic-filters#subjectmatches-mediaarg) (`matches-media` has no regex implementation)

**Reply by gorhill** [reply]
`matches-property` is meant to match properties of elements, `matches-media` is something else.

**Comment by gorhill** [comment, marked as answer]
> Does uBlock support it, or it may support it in the future?

It's not supported. It may be supported if maintainers ask for it with cases. As of now I see only 11 occurrences in [AdGuard Base](https://raw.githubusercontent.com/AdguardTeam/FiltersRegistry/master/filters/filter_2_Base/filter.txt).

**Comment by superlex** [comment]
Ok, so if these rules don't cause issues to uBO we can leave things as they are.
Thanks :)

**Comment by krystian3w** [comment]
This is same or different:

https://github.com/gorhill/uBlock/commit/aca7674bac

Shorter name can glitch unconverted AdGuard syntax.

**Reply by garry-ut99** [reply]
Maybe some bug,  an example filter with a random non-existing property name:
`github.com##div:matches-prop(g43t34ty)` 
still highlights all div elements, just like `github.com##div`
It shouldn't highlight anything because the random property name doesn't exist.

**Reply by gorhill** [reply]
> Shorter name can glitch unconverted AdGuard syntax.

It's not fully compatible with AdGuard's `:matches-property()` as `:matches-prop()` cannot have the property name part being a regex. Supporting and using a regex as property name would require the operator to loop over all properties of an object, which would probably kill performance to the point of being unusable.

**Reply by gorhill** [reply]
> It shouldn't highlight anything because the random property name doesn't exist.

Fixed with <https://github.com/gorhill/uBlock/commit/47b985a0563d8e75420ff89de65c6ab45fb0b01b>

**Reply by garry-ut99** [reply]
I didn't investigate so much, so I'm not sure whether this is a bug or not, and whether it's an issue on browser's side or on uBO's side, but there seems to be a problem with handling `space` character _literally_ when it's placed at the beginning of the property value, it occurs on both Chormium and Firefox, for example, one of buttons (named "master" on top-left) on `https://github.com/gorhill/uBlock/commits/master/` has a property displayed in browser dev console as: `innerText: " master"`, then trying in Element Picker:
- `##button:matches-prop(innerText= master)` doesn't work
- `##button:matches-prop(innerText=/ master/)` doesn't work
- `##button:matches-prop(innerText=/\smaster/)` works
- `##button:matches-prop(innerHTML=/^<span data/)` works

(Other explanation (but I'm not sure about that as I didn't investigate so much) might be that:
- `space` char in browser just looks like `space`, but in practice is some other character than `space` and gets malformed during copy&pasting from browser to Picker and gets turned into `space` in Picker,
- or leading `space` char is just removed by browsers internally but still displayed in dev console, but then why
`##button:matches-prop(innerText=master)` doesn't work)

**Reply by gorhill** [reply]
If value is not a regex, property value has to be an exact match.

![image](https://github.com/uBlockOrigin/uBlock-issues/assets/585534/d18324e4-c0ed-4aa6-8211-fa40cf99bcfb)

**Reply by garry-ut99** [reply]
I used the instructions from: https://github.com/AdguardTeam/ExtendedCss?tab=readme-ov-file#extended-css-matches-property :

> For filters maintainers: To check properties of a specific element, do the following:
> 
> Inspect the page element or select it in Elements tab of browser DevTools.
> Run console.dir($0) in Console tab.

The result in Chromium and Firefox is " master" and nothing more, hence `##button:matches-prop(innerText= master)` is an exact match:
![ss](https://github.com/uBlockOrigin/uBlock-issues/assets/72945564/8f71e842-3a6f-4dc8-a5ec-a40f788131e5)

Even this regex doesn't work: `##button:matches-prop(innerText=/ master/)`

Unless I'm missing something.

**Reply by garry-ut99** [reply]
@iam-py-test if you don't take actual place in discussions nor you don't create comments / nor you don't help / nor you don't provide answers / solutions nor you don't make any contribution to a discussion, then gtfo, and go troll elsewhere with your thumb up emojis and stop sticking them to your idol (gorhill) to help him beating his "opponents" in discussion", stop licking his arsehole each time he replies, because it is neither a porn tube nor facebook, you are useless in our discussions, it's not the first time you are spamming with your emojis, you've been doing it for a long time, dozen of times, in dozen of issues/discussions, go elsewhere.

**Reply by gorhill** [reply]
Use quotes when you want to mind leadeing/trailing spaces:

    ##button:matches-prop(innerText=" master")

Note that I used copy/paste. There is a non-breakable space in there, a plain space won't work.

---

Actually quotes are not needed, just need to use the non-breakable space as this is what is used in the property value.

**Reply by peace2000** [reply]
For Firefox users there's a convenient way to check properties of an element: right click an element in dev tools and click "Show DOM properties".

**Reply by garry-ut99** [reply]
> gorhill : Use quotes when you want to mind leadeing/trailing spaces:
> gorhill : Actually quotes are not needed

I know, I was aware of quotes, I already tried quotes before I created my [comment](https://github.com/uBlockOrigin/uBlock-issues/discussions/2981#discussioncomment-9831456) here about the issue, but:
- since quotes made no difference 
- and also because they are not required by the scriptlet
- and in order not to go unnecessarily beyond the specification
 
 I intentionally didn't include them in my examples I provided.

> gorhill : Note that I used copy/paste. 

Me too, like I mentioned earlier, which caused the issue.

> gorhill : just need to use the non-breakable space as this is what is used in the property value.
> gorhill : There is a non-breakable space in there, a plain space won't work.

Yes, this seems to be it, copy/pasting the value directly from browser console in both Firefox & Chromium resulted in a [normal space](https://unicode-explorer.com/c/0020) `U+0020` instead of [non-breaking space](https://unicode-explorer.com/c/00A0) `U+00A0`, after replacing in Picker manually a normal space with a non-breaking space, the filter works fine. 

After investigation it has turned out that copying property value from browser console:
-  via context menu "Copy message" results in 0020 (space)
-  via context menu "Copy object" results in 00A0 (non-breaking space)

![Schowek01](https://github.com/uBlockOrigin/uBlock-issues/assets/72945564/2ee1f7bd-f7b2-413b-892e-5b3d10413767)

As for now I don't have idea why it happens because I didn't investigate so far, perhaps there is an explanation somewhere in browser documentation.

> peace2000 : For Firefox users there's a convenient way to check properties of an element: right click an element in dev tools and click "Show DOM properties".

Yes, I know, I was aware of it too, I realised it soon after I tried method from AdGuard link, also worth to note that it doesn't solve the issue with the space, because it acts same way like method from AdGuard link, and the space is still copy&pasted as a normal space via "Copy message", to solve the issue needs to copy via context menu "Copy object".
